BUG fix deprecation of `limit` and `fill_method` in `pct_change`

Charlie-XIAO opened the pull request:
- [ ] Towards #53491
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests)
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit)
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file

I'm thinking that we should not over-complicate things by allowing `obj.pct_change()`. The thing is, even if users do `obj.ffill/bfill().pct_change()`, a warning will still be raised unless we check for NA values, for instance, #54981. I propose that we should let users use `obj.ffill/bfill().pct_change(fill_method=None)` to suppress the warnings, and this PR intends to give extra specific deprecation warning messages.

I haven't added test cases for it, but want to make sure if this is the right way to go. Ping @rhshadrach and @jbrockmendel who were involved in the discussion in the original issue. I'm seeing many users and libraries complaining about this might be better to get this done before the next release.

Charlie-XIAO committed:
extra specific deprecation messages

Charlie-XIAO committed:
deprecation in docstring

Charlie-XIAO committed:
Merge remote-tracking branch 'upstream/main' into redepr-pct-change

Charlie-XIAO committed:
use fill_method=None

rhshadrach commented on pandas/core/generic.py:
```diff
@@ -11817,36 +11818,70 @@ def pct_change(
-        if fill_method is not lib.no_default or limit is not lib.no_default:
+        # GH#53491: deprecate the `fill_method` and `limit` keyword, except
+        # `fill_method=None` that does not fill missing values
+        if fill_method not in (lib.no_default, None) and limit is not lib.no_default:
```
I don't think we need warning messages for every single case. Does something like

> The 'fill_method' being not None and the 'limit' argument are deprecated. Either fill in NA values prior to calling pct_change or specify fill_method=None.

work?

Charlie-XIAO replied:
I think we may want to prompt users to use `fill_method=None` even if they have filled NA values. See the original comment of this PR. The case is, even after `ffill` or `bfill`, calling `pct_change` without keyword will raise deprecation warning unless we explicitly check if there are NA values to fill. However, I don't think this is a good approach: (1) this may add too much overhead, and (2) if a user is not filling NA values and uses `pct_change` without keyword, and if the data occasionally does not contain NA values, he/she will not get a warning message and the logic would be incorrect.

Due to these reasons, I think this deprecation would be especially confusing, especially since we are having "incorrect" deprecation warnings in the current version. That's why I'm trying to give extra specific guide for each case. If maintainers do not think this is necessary, I can implement using only a single message.

Charlie-XIAO replied:
Still, I need a confirmation about whether we should prompt users to do `obj.ffill/bfill().pct_change(fill_method=None)` or `obj.ffill/bfill().pct_change()`. I prefer the former.

rhshadrach replied:
> However, I don't think this is a good approach: (1) this may add too much overhead, and (2) if a user is not filling NA values and uses `pct_change` without keyword, and if the data occasionally does not contain NA values, he/she will not get a warning message and the logic would be incorrect.

I'm seeing the current logic takes 7.5% of the runtime for the current warning on a Series with 100k rows - I don't think overhead is a concern. This will cause users to modify their code unnecessarily in what I think is the uncommon case. They will then need to change their code again when we deprecate the `fill_method` argument. I do not think we should do that.

> The thing is, even if users do obj.ffill/bfill().pct_change(), a warning will still be raised unless we check for NA values

That is why I am suggesting the language "Either fill in NA values prior to calling..." and not saying to use ffill/bfill.

Charlie-XIAO replied:
Sure, I will implement your suggestions.

rhshadrach replied:
Perhaps "Either fill in any non-leading NA values" is better.

phofl replied:
agreed with @rhshadrach

rhshadrach reviewed (changes requested):
Thanks for the PR, I'm guessing tests need to be updated.

Charlie-XIAO committed:
apply richard's suggestions

Charlie-XIAO committed:
Merge remote-tracking branch 'upstream/main' into redepr-pct-change

Charlie-XIAO committed:
update tests correspondingly

Charlie-XIAO commented:
@rhshadrach I've implemented your suggestions, please check if it is correct. I've also updated the test cases and now they seem to pass correctly. Not sure if I need to add some additional tests?

By the way, for instance
```python
>>> ser = pd.Series([np.nan, 1, 2, 3, np.nan])
>>> ser.bfill().pct_change()
```
still raises a warning. Should we fix that?

rhshadrach commented:
/preview

rhshadrach commented on pandas/tests/frame/methods/test_pct_change.py:
```diff
@@ -144,12 +143,15 @@ def test_pct_change_with_duplicated_indices(fill_method):
-    )
-    with tm.assert_produces_warning(FutureWarning, match=msg):
-        result = data.pct_change(fill_method=fill_method)
+    if fill_method is not None:
+        msg = (
+            "The 'fill_method' keyword being not None and the 'limit' keyword in "
+            "DataFrame.pct_change are deprecated"
+        )
+        with tm.assert_produces_warning(FutureWarning, match=msg):
+            result = data.pct_change(fill_method=fill_method)
+    else:
+        result = data.pct_change(fill_method=None)
```
assert_produces_warning (unlike pytest.raises) can accept `None` instead of `FutureWarning`. Can you make this something like:

```
warn = None if fill_method is None else FutureWarning
with tm.assert_produces_warning(warn, match=msg):
   ...
```

Charlie-XIAO replied:
Amazing feature, never noticed that before.

rhshadrach commented on pandas/tests/series/methods/test_pct_change.py:
```diff
@@ -101,9 +100,15 @@ def test_pct_change_with_duplicated_indices(fill_method):
-    msg = "The 'fill_method' and 'limit' keywords in Series.pct_change are deprecated"
-    with tm.assert_produces_warning(FutureWarning, match=msg):
-        result = s.pct_change(fill_method=fill_method)
+    if fill_method is not None:
+        msg = (
+            "The 'fill_method' keyword being not None and the 'limit' keyword in "
+            "Series.pct_change are deprecated"
+        )
+        with tm.assert_produces_warning(FutureWarning, match=msg):
+            result = s.pct_change(fill_method=fill_method)
+    else:
+        result = s.pct_change()
```
Ditto.

rhshadrach reviewed (changes requested):
Looks good! Small request on the tests.

rhshadrach commented:
Also - this needs a whatsnew note.

Charlie-XIAO committed:
Merge remote-tracking branch 'upstream/main' into redepr-pct-change

Charlie-XIAO committed:
changelog added; suggestions from Richard

Charlie-XIAO commented:
Done, not sure if the changelog should be added in `v2.2.0` or in `v2.1.2`. Currently it is in `v2.2.0` but let me know if this is wrong or the wording needs to be corrected.

Also, may I have your response to https://github.com/pandas-dev/pandas/pull/55527#issuecomment-1763606520 please @rhshadrach?

rhshadrach commented:
From https://github.com/pandas-dev/pandas/pull/55527#issuecomment-1763606520:

> still raises a warning. Should we fix that?

I believe the behavior is going to change between now and pandas 3.0. So that means we need to warn.

rhshadrach commented on doc/source/whatsnew/v2.2.0.rst:
```diff
@@ -269,7 +269,7 @@ Other Deprecations
 - Deprecated the previous implementation of :class:`DataFrame.stack`; specify ``future_stack=True`` to adopt the future version (:issue:`53515`)
 - Deprecating downcasting the results of :meth:`DataFrame.fillna`, :meth:`Series.fillna`, :meth:`DataFrame.ffill`, :meth:`Series.ffill`, :meth:`DataFrame.bfill`, :meth:`Series.bfill` in object-dtype cases. To opt in to the future version, use ``pd.set_option("future.no_silent_downcasting", True)`` (:issue:`54261`)
--
+- Reverted deprecation of ``fill_method=None`` in :meth:`DataFrame.pct_change`, :meth:`Series.pct_change`, :meth:`DataFrameGroupBy.pct_change`, and :meth:`SeriesGroupBy.pct_change`. Now the keyword ``fill_method`` not being ``None`` and the limit keyword are deprecated. (:issue:`53491`)
```
```suggestion
- Reverted deprecation of ``fill_method=None`` in :meth:`DataFrame.pct_change`, :meth:`Series.pct_change`, :meth:`DataFrameGroupBy.pct_change`, and :meth:`SeriesGroupBy.pct_change`; the values `'backfill'`, `'bfill'`, `'pad'`, and `'ffill'` are still deprecated (:issue:`53491`)
```

I'd like to make it a little more clear this isn't introducing any new deprecations.

rhshadrach replied:
Ah - and as you asked above, yes, move this to 2.1.2.

rhshadrach reviewed (changes requested):
Missed one request in the whatsnew, otherwise I think we're all set.

rhshadrach reviewed (changes requested):
Missed one request in the whatsnew, otherwise I think we're all set.

Charlie-XIAO committed:
Merge remote-tracking branch 'upstream/main' into redepr-pct-change

Charlie-XIAO committed:
updated changelog

Charlie-XIAO commented:
Done @rhshadrach, thanks for your review!

lithomas1 commented:
Bumping off the milestone.

rhshadrach reviewed (approved):
lgtm

rhshadrach commented on doc/source/whatsnew/v2.1.2.rst:
```diff
@@ -8,6 +8,14 @@ including other versions of pandas.
 {{ header }}
 
+.. ---------------------------------------------------------------------------
+.. _whatsnew_220.deprecations:
```
Needs to be 212 now.

rhshadrach replied:
I took care of this to get it into 2.1.2

rhshadrach committed:
Update v2.1.2.rst

rhshadrach reviewed (approved):
lgtm

rhshadrach commented:
Thanks @Charlie-XIAO

rhshadrach commented:
@meeseeksdev backport 2.1.x
